instrumented/integration/src/androidTest/kotlin/com/datadog/android/sdk/integration/rum/ResourceTrackingTest.kt

```diff
@@ -0,0 +1,147 @@
+
+    private lateinit var okHttpClient: OkHttpClient
+    private lateinit var extraAttributes: Map<String, Any?>
+
+    @Before
+    fun setUp() {
+
+        extraAttributes = forge.exhaustiveAttributes()
+
+        okHttpClient = OkHttpClient.Builder()
+            .addInterceptor(
+                RumInterceptor(
+                    rumResourceAttributesProvider = object : RumResourceAttributesProvider {
+                        override fun onProvideAttributes(
+                            request: Request,
+                            response: Response?,
+                            throwable: Throwable?
+                        ): Map<String, Any?> = extraAttributes
+                    }
+                )
+            )
+            .build()
+    }
+
+    @Test
+    fun verifyAttributesAreSentWhenRequestIsSuccessful() {
+
+        val resourceUrl = "${mockServerRule.getConnectionUrl()}/nyan-cat.gif"
+
+        okHttpClient.newCall(
+            Request.Builder().url(resourceUrl).build()
+        ).execute()
+
+        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
+        Thread.sleep(RumTest.FINAL_WAIT_MS)
+
+        val resourceRequests =
+            mockServerRule.getRequests()
+                .rumEventsBy { it.has("resource") && it.get("type").asString == "resource" }
+
+        resourceRequests.verifyEventMatches(
+            listOf(
+                ExpectedResourceEvent(
+                    url = resourceUrl,
+                    statusCode = 200,
+                    extraAttributes = extraAttributes
+                )
+            )
+        )
+    }
+
+    @Test
+    fun verifyAttributesAreSentWhenRequestHasException() {
+
+        val resourceUrl = mockServerRule.getConnectionUrl() +
+            MockServerActivityTestRule.CONNECTION_ISSUE_PATH
+
+        assertThrows(IOException::class.java) {
+            okHttpClient.newCall(
+                Request.Builder().url(resourceUrl).build()
+            ).execute()
+        }
+
+        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
+        Thread.sleep(RumTest.FINAL_WAIT_MS)
+
+        val resourceRequests =
+            mockServerRule.getRequests()
+                .rumEventsBy { it.has("error") && it.get("type").asString == "error" }
+
+        resourceRequests.verifyEventMatches(
+            listOf(
+                ExpectedErrorEvent(
+                    url = resourceUrl,
+                    extraAttributes = extraAttributes,
+                    isCrash = false,
+                    source = ErrorSource.NETWORK
+                )
+            )
+        )
+    }
+
+    private fun List<HandledRequest>.rumEventsBy(
+        predicate: (rawEvent: JsonObject) -> Boolean
+    ): List<JsonObject> {
+        return filter { it.url?.isRumUrl() ?: false }
+            .flatMap { rumPayloadToJsonList(it.textBody!!) }
+            .filter { predicate.invoke(it) }
+    }
+
+    private fun Forge.exhaustiveAttributes(): Map<String, Any?> {
```

Comment:
maybe we should extract this in a `ForgeExt.kt` file. I am pretty sure we will need it in some other places also.